engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/sink/AvroSinkValueParameter.scala

```diff
@@ -0,0 +1,83 @@
+import pl.touk.nussknacker.engine.avro.KafkaAvroBaseTransformer.{SchemaVersionParamName, SinkKeyParamName, SinkValidationModeParameterName, SinkValueParamName, TopicParamName}
+import pl.touk.nussknacker.engine.avro.sink.AvroSinkValueParameter.FieldName
+import pl.touk.nussknacker.engine.definition.parameter.editor.ParameterTypeEditorDeterminer
+
+private[sink] case object AvroSinkValueParameter {
+
+  type FieldName = String
+
+  private val restrictedParamNames = Set(SchemaVersionParamName, SinkKeyParamName, SinkValidationModeParameterName, TopicParamName)
+
+  def apply(typing: TypingResult)(implicit nodeId: NodeId): Validated[ProcessCompilationError, AvroSinkValueParameter] =
+    toSinkValueParameter(typing, paramName = None, isTopLevel = true)
+
+  private def toSinkValueParameter(typing: TypingResult, paramName: Option[String], isTopLevel: Boolean)
+                                  (implicit nodeId: NodeId): Validated[ProcessCompilationError, AvroSinkValueParameter] =
+    typing match {
+      case typed.typing.Unknown =>
+        Invalid(CustomNodeError(nodeId.id, "Cannot determine typing for provided schema", None))
+
+      case typedObject: TypedObjectTypingResult if containsRestrictedNames(typedObject) =>
+        Invalid(CustomNodeError(nodeId.id, s"""Record field name is restricted. Restricted names are ${restrictedParamNames.mkString(", ")}""", None))
+
+      /* kafka-avro-serializer does not support Array at top level
+         [https://github.com/confluentinc/schema-registry/issues/1298] */
+      case TypedClass(clazz, _) if isTopLevel && clazz == classOf[java.util.List[_]] =>
+        Invalid(CustomNodeError(nodeId.id, "Unsupported Avro type. Top level Arrays are not supported", None))
+
+      case typedObject: TypedObjectTypingResult =>
+        val listOfValidatedFieldParams = typedObject.fields.map { case (fieldName, typing) =>
+          val concatName = paramName.map(pn => s"$pn.$fieldName").getOrElse(fieldName)
+          (fieldName, toSinkValueParameter(typing, Some(concatName), isTopLevel = false))
+        }.toList
+        sequence(listOfValidatedFieldParams).map(l => AvroSinkRecordParameter(l.toMap))
+
+      case _ =>
+        val parameter = Parameter(paramName.getOrElse(SinkValueParamName), typing)
+          .copy(
+            isLazyParameter = true,
+            editor = new ParameterTypeEditorDeterminer(typing).determine())
+        Valid(AvroSinkPrimitiveValueParameter(parameter))
+    }
+
+  private def containsRestrictedNames(obj: TypedObjectTypingResult): Boolean = {
+    val fieldNames = obj.fields.keySet
+    fieldNames.nonEmpty & (fieldNames & restrictedParamNames).nonEmpty
+  }
+
+  private def sequence(l: List[(FieldName, Validated[ProcessCompilationError, AvroSinkValueParameter])])
+  : Validated[ProcessCompilationError, List[(FieldName, AvroSinkValueParameter)]] = {
+    val zero: Validated[ProcessCompilationError, List[(FieldName, AvroSinkValueParameter)]] = Valid(Nil)
```

Comment:
I think this method shows we should use ValidatedNel and applicative here, now we're returning only first error?

Reply:
done here: 
https://github.com/TouK/nussknacker/pull/1405/commits/f15c331bf3cdedc082358da86a6a7b9027f12e02